### PR TITLE
feat:(thrift) nested thrift type constructor API

### PR DESCRIPTION
**Pull request by AsterDY**
#### What type of PR is this?

feat
#### Check the PR title.

- [ ] This PR title match the format: \<type\>(optional scope): \<description\>
- [ ] The description of this PR title is user-oriented and clear enough for others to understand.
- [ ] Attach the PR updating the user documentation if the current PR requires user awareness at the usage level. [User docs repo](https://github.com/cloudwego/cloudwego.github.io)

#### (Optional) Translate the PR title into Chinese.
1. 提供thrift嵌套类型的构造函数
  - NewNodeList
  - NewNodeSet
  - NewNodeMap
  - NewNodeStruct
  - NewNodeAny
2. 修复 `NewTypedNode` 未设置实际内容的问题

#### (Optional) More detailed description for this PR(en: English/zh: Chinese).

en:
zh(optional):

### CODE REVIEW

File: thrift/binary.go [hunk 1]
```diff
@@ -1292,6 +1293,362 @@ func (p *BinaryProtocol) DecodeText(val string, desc *TypeDescriptor, disallowUn
...
+	case reflect.Slice:
+		return LIST
+	case reflect.Map:
+		return MAP
+	case reflect.Struct:
+		return STRUCT
+	case reflect.Ptr:
+		return GoType2ThriftType(reflect.ValueOf(val).Elem().Interface())
+	default:
+		return STOP
```
[the remark this answers is not in the document]
Comment (AsterDY) [reply thrift/binary.go]: change to return error

File: thrift/binary.go [hunk 2]
```diff
@@ -1292,6 +1293,394 @@ func (p *BinaryProtocol) DecodeText(val string, desc *TypeDescriptor, disallowUn
...
+		if strAsBinary {
+			return p.ReadBinary(false)
+		}
+		return p.ReadString(false)
+	case LIST, SET:
+		elemType, size, e := p.ReadListBegin()
+		if e != nil {
+			return nil, e
+		}
+		ret := make([]interface{}, 0, size)
```
Comment (felix021) [inline thrift/binary.go:1380]: Vulnerable to maliciously constructed binary with an unreasonable large size. Same below.

File: thrift/binary.go [hunk 3]
```diff
@@ -1292,6 +1293,394 @@ func (p *BinaryProtocol) DecodeText(val string, desc *TypeDescriptor, disallowUn
...
+			}
+			e = p.WriteSetBegin(et, len(v))
+			if e != nil {
+				return 0, e
+			}
+		} else {
+			et, e := GoType2ThriftType(v[0])
+			if e != nil {
+				return 0, e
+			}
```
Comment (felix021) [inline thrift/binary.go:1517]: duplicated with line 1505-1508. How about moving them before `if`?

File: thrift/binary.go [hunk 4]
```diff
@@ -1292,6 +1293,394 @@ func (p *BinaryProtocol) DecodeText(val string, desc *TypeDescriptor, disallowUn
...
+			e = p.WriteListBegin(et, len(v))
+			if e != nil {
+				return 0, e
+			}
```
Comment (felix021) [inline thrift/binary.go:1521]: Same here (1515-1517), moving after `if-else`?

File: thrift/binary.go (hunk 2 above)
Comment (AsterDY) [reply thrift/binary.go:1380]: this is hard to prevent..

File: thrift/binary.go (hunk 4 above)
Comment (AsterDY) [reply thrift/binary.go:1521]: 严格来说不应该，但是实际上也不是不可以

File: thrift/binary.go [hunk 5]
```diff
@@ -1292,6 +1293,394 @@ func (p *BinaryProtocol) DecodeText(val string, desc *TypeDescriptor, disallowUn
...
+			if e := p.WriteFieldBegin("", ft, k); e != nil {
+				return 0, e
+			}
+			if _, e := p.WriteAny(vv, sliceAsSet); e != nil {
+				return 0, e
+			}
+		}
+		return STRUCT, p.WriteFieldStop()
+	default:
```
Comment (felix021) [inline thrift/binary.go:1679]: Suggestion: opensource task for the community to encode arbitrary type with reflection, which supports zero-length slice/map.

Comment (felix021) [reply thrift/binary.go:1679]: Another approach may be to generate all possible combinations of the types based on `go generate` with a template to achieve best performance.

File: thrift/generic/node.go [hunk 6]
```diff
@@ -157,30 +143,112 @@ func NewNodeBinary(val []byte) Node {
 	return NewNode(thrift.STRING, buf)
 }
 
+// NewNodeList creates a LIST node.
+// The element thrift type depends on vals' concrete type,
+// thus there must be at least one val.
+func NewNodeList(vals []interface{}) Node {
+	p := thrift.NewBinaryProtocol(make([]byte, 0, len(vals)*64))
```
Comment (felix021) [inline thrift/generic/node.go:150]: How are those constants chosen? Global constant suggested btw.

File: thrift/generic/node.go [hunk 7]
```diff
@@ -157,30 +143,112 @@ func NewNodeBinary(val []byte) Node {
...
+	}
+	return NewNode(thrift.SET, p.Buf)
+}
+
+// NewNodeMap creates a MAP node.
+// The thrift type of key and element depends on kvs' concrete type,
+// thus there must be at least one kv.
+func NewNodeMap(kvs map[interface{}]interface{}) Node {
+	p := thrift.NewBinaryProtocol(make([]byte, 0, len(kvs)*128))
+	if _, err := p.WriteAny(kvs, false); err != nil {
```
Comment (felix021) [inline thrift/generic/node.go:173]: Why it's `false` here? what if the key or value is a set?

Comment (felix021) [reply thrift/generic/node.go:173]: Same below.

File: thrift/binary.go (hunk 5 above)
Comment (AsterDY) [reply thrift/binary.go:1679]: TODO

File: thrift/generic/node.go (hunk 6 above)
Comment (AsterDY) [reply thrift/generic/node.go:150]: Experience. 我加一下吧

File: thrift/generic/node.go (hunk 7 above)
Comment (AsterDY) [reply thrift/generic/node.go:173]: 加个选项吧

**Review by felix021 — APPROVED** [review]

**Pull request merged**